cardano-api/src/Cardano/Api/IPC/Monad.hs

```diff
@@ -1,70 +1,164 @@
 {-# LANGUAGE DerivingVia #-}
+{-# LANGUAGE GADTs #-}
 {-# LANGUAGE GeneralisedNewtypeDeriving #-}
 
 module Cardano.Api.IPC.Monad
-  ( LocalStateQueryScript
-  , sendMsgQuery
-  , setupLocalStateQueryScript
+  ( LocalStateQueryExpr
+  , executeQueryLocalState
+  , executeQueryLocalStateWithChainSync
+  , queryExpr
+  , determineEraExpr
   ) where
 
 import Cardano.Api.Block
+import Cardano.Api.Eras
 import Cardano.Api.IPC
+import Cardano.Api.Modes
 import Control.Applicative
 import Control.Concurrent.STM
 import Control.Monad
 import Control.Monad.IO.Class
 import Control.Monad.Trans.Cont
+import Control.Monad.Trans.Except
+import Data.Bifunctor
 import Data.Either
 import Data.Function
 import Data.Maybe
-import Data.Ord
 import Shelley.Spec.Ledger.Scripts ()
 import System.IO
 
+import qualified Ouroboros.Network.Protocol.ChainSync.Client as Net.Sync
 import qualified Ouroboros.Network.Protocol.LocalStateQuery.Client as Net.Query
 import qualified Ouroboros.Network.Protocol.LocalStateQuery.Type as Net.Query
 
--- | Monadic type for constructing local state queries.
+{- HLINT ignore "Use let" -}
+
+-- | Monadic type for constructing local state query expressions.
+--
+-- Use 'queryExpr' in a do block to construct queries of this type and convert
+-- the expression to a 'Net.Query.LocalStateQueryClient' with 'setupLocalStateQueryExpr'.
+--
+-- Some consideration was made to use Applicative instead of Monad as the abstraction in
+-- order to support pipelining, but we actually have a fair amount of code where the next
+-- query depends on the result of the former and therefore actually need Monad.
 --
--- Use 'sendMsgQuery' in a do block to construct queries of this type and convert
--- the expression to a 'Net.Query.LocalStateQueryClient' with 'setupLocalStateQueryScript'.
-newtype LocalStateQueryScript block point query r m a = LocalStateQueryScript
-  { runLocalStateQueryScript :: ContT (Net.Query.ClientStAcquired block point query m r) m a
+-- In order to make pipelining still possible we can explore the use of Selective Functors
+-- which would allow us to straddle both worlds.
+newtype LocalStateQueryExpr block point query r m a = LocalStateQueryExpr
+  { runLocalStateQueryExpr :: ContT (Net.Query.ClientStAcquired block point query m r) m a
   } deriving (Functor, Applicative, Monad, MonadIO)
 
--- | Use 'sendMsgQuery' in a do block to construct monadic local state queries.
-sendMsgQuery :: Monad m => query a -> LocalStateQueryScript block point query r m a
-sendMsgQuery q = LocalStateQueryScript . ContT $ \f -> pure $
-  Net.Query.SendMsgQuery q $
-    Net.Query.ClientStQuerying
-    { Net.Query.recvMsgResult = f
-    }
+-- | Execute a local state query expression.
+executeQueryLocalState
+  :: LocalNodeConnectInfo mode
+  -> Maybe ChainPoint
+  -> (AcquireFailure -> e)
+  -> (NodeToClientVersion -> ExceptT e (LocalStateQueryExpr (BlockInMode mode) ChainPoint (QueryInMode mode) () IO) a)
+  -> IO (Either e a)
```

Comment:
Please add documentation for the `Maybe ChainPoint` argument.

This seems overly complicated. Why not just:

```haskell
executeQueryLocalState
  :: LocalNodeConnectInfo mode
  -> Maybe ChainPoint
  -> (NodeToClientVersion -> LocalStateQueryExpr (BlockInMode mode) ChainPoint (QueryInMode mode) () IO a)
  -> IO (Either AcquireFailure a)
```

Reply:
I'm trying to find the right ergonomics for the use site.  A lot of the time these monadic expressions will want error handling and I'm trying to avoid multilayered Eithers and Maybes as much as possible.

Reply:
`Maybe ChainPoint` was pre-existing.  I don't actually know what it is for.